src/foam/nanos/crunch/crunchtest/PredicatedPCJDAOTest.java

```diff
@@ -0,0 +1,146 @@
+      EQ(CapabilityCapabilityJunction.SOURCE_ID, "cap"),
+      EQ(CapabilityCapabilityJunction.TARGET_ID, "nanoPrereq")));
+    test(nanoFindCTP != null, "nanoUser can see ccj[cap, prereq]");
+    test(nanoFindCTTP == null, "nanoUser cannot see ccj[cap, testPrereq]");
+    test(nanoFindCTNP != null, "nanoUser can see ccj[cap, nanoPrereq]");
+    // test nanoAdmin access
+    CapabilityCapabilityJunction adminFindCTP = (CapabilityCapabilityJunction) prerequisiteCapabilityJunctionDAO.inX(adminX).find(AND(
+      EQ(CapabilityCapabilityJunction.SOURCE_ID, "cap"),
+      EQ(CapabilityCapabilityJunction.TARGET_ID, "prereq")));
+    CapabilityCapabilityJunction adminFindCTTP = (CapabilityCapabilityJunction) prerequisiteCapabilityJunctionDAO.inX(adminX).find(AND(
+      EQ(CapabilityCapabilityJunction.SOURCE_ID, "cap"),
+      EQ(CapabilityCapabilityJunction.TARGET_ID, "testPrereq")));
+    CapabilityCapabilityJunction adminFindCTNP = (CapabilityCapabilityJunction) prerequisiteCapabilityJunctionDAO.inX(adminX).find(AND(
+      EQ(CapabilityCapabilityJunction.SOURCE_ID, "cap"),
+      EQ(CapabilityCapabilityJunction.TARGET_ID, "nanoPrereq")));
+    test(adminFindCTP != null, "nanoAdmin can see ccj[cap, prereq]");
+    test(adminFindCTTP != null, "nanoAdmin can see ccj[cap, testPrereq]");
+    test(adminFindCTNP != null, "nanoAdmin can see ccj[cap, nanoPrereq]");
+  }
+
+  public void testPredicatedCCJSelect() {
+    // test nanoUser access
+    List<CapabilityCapabilityJunction> nanoSelect = ((ArraySink) prerequisiteCapabilityJunctionDAO.inX(nanoX).where(EQ(CapabilityCapabilityJunction.SOURCE_ID, "cap")).select(new ArraySink())).getArray();
+    test(nanoSelect.size() == 2, "nanoUser select yields 2 prereqs: " + nanoSelect);
+    // test admin access
+    List<CapabilityCapabilityJunction> adminSelect = ((ArraySink) prerequisiteCapabilityJunctionDAO.inX(adminX).where(EQ(CapabilityCapabilityJunction.SOURCE_ID, "cap")).select(new ArraySink())).getArray();
+    test(adminSelect.size() == 3, "nanoAdmin select yields 3 prereqs: " + adminSelect);
+  }
+
+  public void testUserUCJGranting(X x) {
+    UserCapabilityJunction n_ucjP = new UserCapabilityJunction.Builder(x).setSourceId(nanoUser.getId()).setTargetId(prereq.getId()).build();
+    UserCapabilityJunction n_ucjTP = new UserCapabilityJunction.Builder(x).setSourceId(nanoUser.getId()).setTargetId(testPrereq.getId()).build();
+    UserCapabilityJunction n_ucjNP = new UserCapabilityJunction.Builder(x).setSourceId(nanoUser.getId()).setTargetId(nanoPrereq.getId()).build();
+    UserCapabilityJunction n_ucjC = new UserCapabilityJunction.Builder(x).setSourceId(nanoUser.getId()).setTargetId(cap.getId()).build();
+    n_ucjP = (UserCapabilityJunction) userCapabilityJunctionDAO.inX(nanoX).put(n_ucjP);
+    n_ucjTP = (UserCapabilityJunction) userCapabilityJunctionDAO.inX(nanoX).put(n_ucjTP);
+    n_ucjC = (UserCapabilityJunction) userCapabilityJunctionDAO.inX(nanoX).put(n_ucjC);
```

Comment:
A comment before each of these test sections would be useful to explain what is being tested for. e.g 
```suggestion
    n_ucjC = (UserCapabilityJunction) userCapabilityJunctionDAO.inX(nanoX).put(n_ucjC);
    // Testing when all prerequisites of "cap" are granted except for the capability satisfied by the capabilityCapabilityPredicate (user.isSpid=nanopay) between capability "cap" and capability "nanoPrereq". Results in "cap" not being GRANTED.
```

Or provide more info in the description argument of the test.

Reply:
addressed: https://github.com/foam-framework/foam2/pull/4950/commits/c2a9aa9805872ff2390135be46d1b7d5250b310c